tests/Operation/ChangeStreamFunctionalTest.php

```diff
@@ -0,0 +1,252 @@
+
+        $result = $this->collection->insertOne(['x' => 1]);
+        $this->assertInstanceOf('MongoDB\InsertOneResult', $result);
+        $this->assertSame(1, $result->getInsertedCount());
+
+        $changeStreamResult->next();
+        $this->assertSame(1, $changeStreamResult->key());
+
+        $changeStreamResult->next();
+        $this->assertSame(null, $changeStreamResult->key());
+        $changeStreamResult->next();
+        $this->assertSame(null, $changeStreamResult->key());
+
+        $operation = new DatabaseCommand($this->getDatabaseName(), ["killCursors" => $this->getCollectionName(), "cursors" => [$changeStreamResult->getId()]]);
+        $operation->execute($this->getPrimaryServer());
+
+        $changeStreamResult->next();
+        $this->assertSame(null, $changeStreamResult->key());
+
+        $result = $this->collection->insertOne(['x' => 2]);
+        $this->assertInstanceOf('MongoDB\InsertOneResult', $result);
+        $this->assertSame(1, $result->getInsertedCount());
+
+        $changeStreamResult->next();
+        $this->assertSame(2, $changeStreamResult->key());
+    }
+
+    public function testNonEmptyPipeline()
+    {
+        $this->collection = new Collection($this->manager, $this->getDatabaseName(), $this->getCollectionName());
+
+        $pipeline = [['$project' => ['foo' => [0]]]];
+        $changeStreamResult = $this->collection->watch($pipeline, []);
+
+        $result = $this->collection->insertOne(['x' => 1]);
+        $this->assertInstanceOf('MongoDB\InsertOneResult', $result);
+        $this->assertSame(1, $result->getInsertedCount());
+
+        $changeStreamResult->next();
+        $expectedResult = (object) ([
+                            '_id' => $changeStreamResult->current()->_id,
+                            'foo' => [0]
+                        ]);
+        $this->assertEquals($changeStreamResult->current(), $expectedResult);
+    }
+
+    public function testCursorWithEmptyBatchNotClosed()
+    {
+        $this->collection = new Collection($this->manager, $this->getDatabaseName(), $this->getCollectionName());
+
+        $changeStreamResult = $this->collection->watch();
+        $this->assertNotNull($changeStreamResult);
+    }
+
+    /**
+     * @expectedException MongoDB\Exception\ResumeTokenException
+     */
+    public function testFailureAfterResumeTokenRemoved()
+    {
+        $this->collection = new Collection($this->manager, $this->getDatabaseName(), $this->getCollectionName());
+
+        $pipeline =  [['$project' => ['_id' => 0 ]]];
+        $changeStreamResult = $this->collection->watch($pipeline, []);
+
+        $result = $this->collection->insertOne(['x' => 1]);
+        $this->assertInstanceOf('MongoDB\InsertOneResult', $result);
+        $this->assertSame(1, $result->getInsertedCount());
+
+        $changeStreamResult->next();
+    }
+
+    public function testConnectionException()
+    {
+        $client = new Client($this->getUri(), ['socketTimeoutMS' => 1005], []);
```

Comment:
If this sporadically fails, that may be a question for a server engineer. I would expect the `aggregate` command to immediately return either an empty tailable cursor or one with a few results (if you provided a resume token).

1005 might still be higher than needed if the default `maxAwaitTimeMS` is one second.

If there is no middle ground here (and `aggregate` does block), we might have to wait until `maxAwaitTimeMS` is supported in PHPC to test this, as we can then set our await time to 2 seconds and provide 1500 as a socket timeout and hope that `aggregate` always returns within that time.